### DISCUSSION TITLE
model.copy is useful for frozen=True

### DISCUSSION THREAD

**Discussion by JacobHayes** [pydantic/pydantic#4250] [Pydantic V2]
First off, I'm very excited for pydantic v2! The main thing I'd like to see preserved if possible is the `.copy(update=...)` ability, which is particularly useful when working with frozen models. It seems this was also [requested here](https://github.com/samuelcolvin/pydantic/discussions/3352) (perhaps before `.copy(update=...)` was added?).

I don't much care if the interface changes, but I'm not sure how that'd work with the stdlib's `copy` module. Would there be an alternate way to handle copy-and-update for frozen models, or would the suggestion be to add my own `.copy` method using stdlib `.copy` and then bypassing frozen somehow?

Thanks!

**Comment by PrettyWood** [comment]
@JacobHayes I'm really in favor of dropping the `copy(update=...)`, which is a bad paradigm and has issues (it does not validate things)
IMO we should do like `dataclass`

```py
from dataclasses import dataclass, field
from copy import copy, deepcopy


@dataclass
class M:
    a: int
    b: str
    c: list[int] = field(default_factory=list)

    def copy_with_update(self, **kwargs):
        new_dict = {**self.__dict__, **kwargs}
        return M(**new_dict)

l = [1, 2, 3]
m = M(a=1, b="hello", c=l)
assert repr(m) == "M(a=1, b='hello', c=[1, 2, 3])"

copy_m = copy(m)
assert m.c is copy_m.c

deepcopy_m = deepcopy(m)
assert m.c is not deepcopy_m.c

n = m.copy_with_update(a=2, c=[4, 5, 6])
assert repr(n) == "M(a=2, b='hello', c=[4, 5, 6])"
```

This is why for v2 we'd rather have proper `__copy__` and `__deepcopy__` implementations. If you want to copy and update, you just have to `(deep)copy`, get the dict repr and revalidate or not. It gives you full control of what you want to achieve

**Reply by JacobHayes** [reply]
> it does not validate things

Ah, agreed there! I've [overridden](https://github.com/artigraph/artigraph/blob/golden/src/arti/internal/models.py#L163-L184) `.copy` in my models to add validation (I think with your guidance in other GH issues 😃). Simplifying that will be great.

> If you want to copy and update, you just have to (deep)copy, get the dict repr and revalidate or not.

I think there are two challenges with re-initializing as a way to copy/validate:
- [hard] private attributes will not be copied, as they cannot be passed via `__init__`
    - arguably, I would like to be able to set private attributes in `__init__`, so maybe I should request that? 😉
- [not hard] `__fields_set__` gets updated to have all fields (including defaulted ones)
    - I only care because I [use `__fields_set__` to filter `__repr_args__`](https://github.com/artigraph/artigraph/blob/7cacc870b375237502d021695313922196bc0d27/src/arti/internal/models.py#L149-L150), which really improves readability of deeply nested models.

Working around the `__fields_set__` is easy enough:

```python
    def copy_with_update(self, **kwargs):
        new_dict = {**self.__dict__, **kwargs}
        copy = M(**new_dict)
        object.__setattr__(copy, "__fields_set__", self.__fields_set__ | set(kwargs))
        return copy
```
but copying `PrivateAttr`s is a bit more tricky (have to look them all up, loop, and setattr).

**Reply by samuelcolvin** [reply]
we need to consider how private attributes work in V2 more generally. I don't yet have a good idea of where they should fit in.

**Comment by spewil** [comment]
Sorry to resurrect this, but I'm interested in whether validated copying for frozen models was ever pursued. It seems like a very useful pattern, but I'm happy to be convinced otherwise. I too have implemented my own copy methods which seems redundant. Thanks!
